src/cloud_what/providers/azure.py

```diff
@@ -107,12 +108,16 @@ def is_likely_running_on_cloud(self):
         # We know that Azure uses only HyperV
         if 'virt.host_type' in self.hw_info:
             if 'hyperv' in self.hw_info['virt.host_type']:
-                probability += 0.3
+                probability += 0.2
```

Comment:
I don't see the reason, why this probability should be decreased.

Reply:
The reason why I changed the probability values of virt.host_type and found_microsoft:
- virt.host_type was at 0.3
- found_microsoft was at 0.3

On a pure hyperV, BOTH checks are true and therefore, hyperV is recognized as azure.Therefore I think, this need to be changed!

Reply:
I can see that `is_likely_running_on_cloud()` can give false positive result, but I think that it is good behavior. I will explain it later. First I would like to understand few things:

1) Did you observe this behavior, when subscription-manager was used? If yes, can you please describe this issue with more details?
2) Or do you use python-cloud-what as Python package and do you call `is_likely_running_on_cloud()` from your Python script/application? Why don't you use method `is_running_on_cloud()`?

Second I would like to clarify few things.

I would recommend to use method `is_running_on_cloud()` for detection if given public cloud is used. This should work, when it doesn't work, then it is big issue and we should fix it.

The method `is_likely_running_on_cloud()` was implemented as fallback, when `is_running_on_cloud()` returns false negative result. Thus it is good that `is_likely_running_on_cloud()` returns false positive result on Hyper-V. Without that we wouldn't be able to detect Azure, when Microsoft change strong sign. Anyway the method `is_likely_running_on_cloud()` should be used only as preliminary check if the VM is running on the given public cloud. When `is_likely_running_on_cloud()` returns some value greater than zero, then `get_metadata()` should be called. When this method returns some metadata, then only in this situation you have strong confidence that the VM is running on given public cloud.

BTW: the `provider.get_cloud_provider()` does all of this described in previous paragraph.

The method `detect_cloud_provider()` from `provider` module has default threshold=0.5, but it can be altered.

Reply:
The error was observed by using the subscription-manager directly. 
Actually, if sub-man recognizes, that it runs on a azure (but actually it's a hyperV instance) it tries to read facts for azure, which fails:

```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/subscription_manager/managercli.py", line 1863, in _do_command
    consumer = service.register(
  File "/usr/lib/python3/dist-packages/rhsmlib/services/register.py", line 85, in register
    facts_dict = self.facts.get_facts()
  File "/usr/lib/python3/dist-packages/subscription_manager/facts.py", line 78, in get_facts
    facts = collector.get_all()
  File "/usr/lib/python3/dist-packages/rhsmlib/facts/all.py", line 41, in get_all
    fact_collector = fact_collector_cls(collected_hw_info=results)
  File "/usr/lib/python3/dist-packages/rhsmlib/facts/cloud_facts.py", line 46, in __init__
    self.cloud_provider = get_cloud_provider(self._collected_hw_info)
  File "/usr/lib/python3/dist-packages/cloud_what/provider.py", line 149, in get_cloud_provider
    metadata = cloud_provider.get_metadata()
  File "/usr/lib/python3/dist-packages/cloud_what/providers/azure.py", line 235, in get_metadata
    return super(AzureCloudProvider, self).get_metadata()
  File "/usr/lib/python3/dist-packages/cloud_what/_base_provider.py", line 474, in get_metadata
    return self._get_metadata_from_server()
  File "/usr/lib/python3/dist-packages/cloud_what/providers/azure.py", line 195, in _get_metadata_from_server
    metadata = super(AzureCloudProvider, self)._get_metadata_from_server()
  File "/usr/lib/python3/dist-packages/cloud_what/_base_provider.py", line 402, in _get_metadata_from_server
    self._cached_metadata = self._get_data_from_server(
  File "/usr/lib/python3/dist-packages/cloud_what/providers/azure.py", line 188, in _get_data_from_server
    return super(AzureCloudProvider, self)._get_data_from_server(data_type, url, headers)
  File "/usr/lib/python3/dist-packages/cloud_what/_base_provider.py", line 387, in _get_data_from_server
    response = self._session.send(prepared_http_req, timeout=self.TIMEOUT)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 529, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPConnectionPool(host='169.254.169.254', port=80): Read timed out. (read timeout=10.0)```

Therefore, it's pretty safe to detect azure by using the dmi.chassis.asset_tag and/or the availability of the azure linux agent. We should really prevent that sub-man has false-positive results. This is possible as described in this PR - even if dmidecode isn't available - which shouldn't be the case in 99,5% of the installations.

Reply:
No. It's a private host with no access to the internet which often occurs in katello / rh satellite / orcharhino scenarios. 

Therefore, sub-man should not think its on a azure environment if the signs are pretty clear, that its not azure :)

Reply:
When you look at the traceback, then I can see that the exception `ReadTimeout` was raised. It means that there is some machine in your local network wit this IP and it is dropping all packets with HTTP/HTTPS requests. We should use more generic exception here:

https://github.com/candlepin/subscription-manager/blob/main/src/cloud_what/_base_provider.py#L386

Can you please try to replace `requests.ConnectionError` with `requests.RequestException`.

Reply:
Well, actually I don't have access to this szenario anymore.  So, maybe this could be fixed but I think this would be another PR. 

As I said, in my eyes sub-man should never try to reach the azure metadata server if it's not a azure environment.

Reply:
Yeah, I will create another PR using `requests.RequestException`.

> As I said, in my eyes sub-man should never try to reach the azure metadata server if it's not a azure environment.

Only one fact (VM is running on Hyper-V) could not lead to the false positive detection of Azure. When hypervisor is Hyper-V, then this fact increase probability only to 0.3 and the threshold is 0.5. Thus there has to be another fact increasing the probability (e.g. `dmi.chassis.manufacturer` is `Microsoft`) In such case the total probability would be 0.6 and subscription-manager would try to contact IMDS server an get metadata.

From my perspective it is desired behavior and I can understand that you don't like it... So, there exists workaround for your case. You can create custom facts rewriting `dmi.chassis.manufacturer` with something like `Redacted` or `MS`:

```json
{
    "dmi.chassis.manufacturer": "Redacted"
}
```

Maybe it is also possible to modify information provided by SM BIOS from Hyper-V hypervisor.

Reply:
> Only one fact (VM is running on Hyper-V) could not lead to the false positive detection of Azure.

Unfortunately, the facts look like this on a VM on pure HyperV:

```
{
  "dmi.chassis.manufacturer": "microsoft",
  "virt.host_type": "hyperv",
  "virt.is_guest": true,
}```

found_microsoft: +0.3
'hyperv' in self.hw_info['virt.host_type']: +0.3
--> 0.6 failure

Reply:
Well, I'm starting to consider using following approach: when we want to gather facts, then only detection using strong signs would be used and fallback to heuristics would be used only for automatic registration. We need false positive detection mostly due to automatic registration to be as reliable as possible. 

When detection using strong signs would not work (due to some unexpected change on metadata server of cloud provider) and not all facts would not be gathered/reported from metadata server, then we could fix this issue in strong detection and systems could be updated using dnf update.

Thus can you please change your PR and do not change existing probability?

Reply:
Thank you very much. Understand your point now. 
I have adapted the PR again. Hope its now ready to be merged.